models/glif_psc.cpp

```diff
@@ -0,0 +1,578 @@
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::glif_psc::init_state_( const Node& proto )
+{
+  const glif_psc& pr = downcast< glif_psc >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::glif_psc::init_buffers_()
+{
+  B_.spikes_.clear();   // includes resize
+  B_.currents_.clear(); // include resize
+  B_.logger_.reset();   // includes resize
+}
+
+void
+nest::glif_psc::calibrate()
+{
+  B_.logger_.init();
+
+  V_.t_ref_remaining_ = 0.0;
+  V_.t_ref_total_ = P_.t_ref_;
+  V_.last_spike_ = 0.0;
+  V_.last_voltage_ = 0.0;
+
+  // post synapse currents
+  const double h = Time::get_resolution().get_ms(); // in ms
+
+  V_.P11_.resize( P_.n_receptors_() );
+  V_.P21_.resize( P_.n_receptors_() );
+  V_.P22_.resize( P_.n_receptors_() );
+  V_.P31_.resize( P_.n_receptors_() );
+  V_.P32_.resize( P_.n_receptors_() );
+
+  S_.y1_.resize( P_.n_receptors_() );
+  S_.y2_.resize( P_.n_receptors_() );
+  V_.PSCInitialValues_.resize( P_.n_receptors_() );
+
+  B_.spikes_.resize( P_.n_receptors_() );
+
+  double Tau_ = P_.C_m_ / P_.G_; // in ms
+  V_.P33_ = std::exp( -h / Tau_ );
+  V_.P30_ = 1 / P_.C_m_ * ( 1 - V_.P33_ ) * Tau_;
+
+  for ( size_t i = 0; i < P_.n_receptors_(); i++ )
+  {
+    // these P are independent
+    V_.P11_[ i ] = V_.P22_[ i ] = std::exp( -h / P_.tau_syn_[ i ] );
+
+    V_.P21_[ i ] = h * V_.P11_[ i ];
+
+    // these are determined according to a numeric stability criterion
+    // input time parameter shall be in ms, capacity in pF
+    V_.P31_[ i ] = propagator_31( P_.tau_syn_[ i ], Tau_, P_.C_m_, h );
+    V_.P32_[ i ] = propagator_32( P_.tau_syn_[ i ], Tau_, P_.C_m_, h );
+
+    V_.PSCInitialValues_[ i ] = 1.0 * numerics::e / P_.tau_syn_[ i ];
+    B_.spikes_[ i ].resize();
+  }
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::glif_psc::update( Time const& origin, const long from, const long to )
+{
+  // glif_lif_r_asc_a
+  const double dt = Time::get_resolution().get_ms();
+
+  // get model type
+  std::string model_str = P_.glif_model_;
+  std::transform( model_str.begin(), model_str.end(), model_str.begin(), ::tolower );
+  if ( nest::glif_psc::model_type_lu.find( model_str ) == nest::glif_psc::model_type_lu.end() )
+  {
+    throw BadProperty( "Bad glif model type string." );
+  }
+  long model_type = nest::glif_psc::model_type_lu[ model_str ];
+
+  double v_old = S_.U_;
+  double spike_component = 0.0;
+  double voltage_component = 0.0;
+  double th_old = S_.threshold_;
+
+  for ( long lag = from; lag < to; ++lag )
+  {
+
+    // update threshold via exact solution of dynamics of spike component of
+    // threshold
+    // exact solution of dynamics of spike component of threshold for glif2/4/5
+    // models with "R"
+    if ( model_type == 2 || model_type == 4 || model_type == 5 )
```

Comment:
I would suggest to introduce a flag `has_theta_s_` that is set for those model variants that have it, so this can be just
```c++
if ( has_theta_spike_ )
```
That is faster and more expressive.